File: Iklimku/WeatherModel.swift
import Foundation

// Model untuk data cuaca saat ini
struct CurrentWeather: Codable {
    let temp: Double
    let feels_like: Double
    let humidity: Int
    let wind_speed: Double
    let weather: [WeatherDescription]
    let dt: TimeInterval
    
    var temperature: String {
        return "\(Int(round(temp)))°"
    }
    
    var feelsLike: String {
        return "\(Int(round(feels_like)))°"
    }
    
    var humidityText: String {
        return "\(humidity)%"
    }
    
    var windSpeed: String {
        return "\(Int(round(wind_speed))) km/h"
    }
    
    var weatherDescription: String {
        return weather.first?.description.capitalized ?? ""
    }
    
    var weatherIcon: String {
        return weather.first?.icon ?? ""
    }
    
    var date: Date {
        return Date(timeIntervalSince1970: dt)
    }
}

// Model untuk deskripsi cuaca
struct WeatherDescription: Codable {
    let id: Int
    let main: String
    let description: String
    let icon: String
}

// Model untuk data cuaca harian
struct DailyWeather: Codable {
    let dt: TimeInterval
    let temp: Temperature
    let weather: [WeatherDescription]
    
    var date: Date {
        return Date(timeIntervalSince1970: dt)
    }
    
    var dayName: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "id_ID")
        formatter.dateFormat = "EEEE"
        return formatter.string(from: date)
    }
    
    var dayShort: String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "id_ID")
        formatter.dateFormat = "EEE"
        return formatter.string(from: date)
    }
    
    var maxTemp: String {
        return "\(Int(round(temp.max)))°"
    }
    
    var minTemp: String {
        return "\(Int(round(temp.min)))°"
    }
    
    var weatherIcon: String {
        return weather.first?.icon ?? ""
    }
}

// Model untuk temperatur
struct Temperature: Codable {
    let day: Double
    let min: Double
    let max: Double
    let night: Double
    let eve: Double
    let morn: Double
}

// Model untuk response API cuaca
struct WeatherResponse: Codable {
    let current: CurrentWeather
    let daily: [DailyWeather]
    let timezone: String
    
    var locationName: String {
        return timezone.replacingOccurrences(of: "_", with: " ")
    }
}

// Model untuk lokasi
struct Location: Codable {
    let name: String
    let lat: Double
    let lon: Double
    let country: String
    let state: String?
    
    var displayName: String {
        if let state = state {
            return "\(name), \(state)"
        }
        return "\(name), \(country)"
    }
} 

File: lib/models/weather_model.dart
class WeatherModel {
  final double temperature;
  final double feelsLike;
  final int humidity;
  final double windSpeed;
  final String description;
  final String icon;
  final DateTime date;
  final String location;

  WeatherModel({
    required this.temperature,
    required this.feelsLike,
    required this.humidity,
    required this.windSpeed,
    required this.description,
    required this.icon,
    required this.date,
    required this.location,
  });

  factory WeatherModel.fromJson(Map<String, dynamic> json, String location) {
    return WeatherModel(
      temperature: json['main']['temp'].toDouble(),
      feelsLike: json['main']['feels_like'].toDouble(),
      humidity: json['main']['humidity'],
      windSpeed: json['wind']['speed'].toDouble(),
      description: json['weather'][0]['description'],
      icon: json['weather'][0]['icon'],
      date: DateTime.fromMillisecondsSinceEpoch(json['dt'] * 1000),
      location: location,
    );
  }

  String get temperatureText => '${temperature.round()}°';
  String get feelsLikeText => '${feelsLike.round()}°';
  String get humidityText => '$humidity%';
  String get windSpeedText => '${windSpeed.round()} km/h';
  String get descriptionText => description.capitalize();
}

class DailyForecast {
  final DateTime date;
  final double maxTemp;
  final double minTemp;
  final String description;
  final String icon;

  DailyForecast({
    required this.date,
    required this.maxTemp,
    required this.minTemp,
    required this.description,
    required this.icon,
  });

  factory DailyForecast.fromJson(Map<String, dynamic> json) {
    return DailyForecast(
      date: DateTime.fromMillisecondsSinceEpoch(json['dt'] * 1000),
      maxTemp: json['temp']['max'].toDouble(),
      minTemp: json['temp']['min'].toDouble(),
      description: json['weather'][0]['description'],
      icon: json['weather'][0]['icon'],
    );
  }

  String get dayName {
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);
    final forecastDate = DateTime(date.year, date.month, date.day);

    if (forecastDate == today) {
      return 'Hari Ini';
    } else if (forecastDate == today.add(const Duration(days: 1))) {
      return 'Besok';
    } else {
      final days = [
        'Senin',
        'Selasa',
        'Rabu',
        'Kamis',
        'Jumat',
        'Sabtu',
        'Minggu',
      ];
      return days[date.weekday - 1];
    }
  }

  String get dayShort {
    final days = ['Sen', 'Sel', 'Rab', 'Kam', 'Jum', 'Sab', 'Min'];
    return days[date.weekday - 1];
  }

  String get maxTempText => '${maxTemp.round()}°';
  String get minTempText => '${minTemp.round()}°';
  String get descriptionText => description.capitalize();
}

class Location {
  final String name;
  final double lat;
  final double lon;
  final String country;
  final String? state;

  Location({
    required this.name,
    required this.lat,
    required this.lon,
    required this.country,
    this.state,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      name: json['name'],
      lat: json['lat'].toDouble(),
      lon: json['lon'].toDouble(),
      country: json['country'],
      state: json['state'],
    );
  }

  String get displayName {
    if (name.isEmpty || name.toLowerCase() == 'unknown' || name == '-') {
      return 'Lokasi tidak tersedia';
    }
    if (state != null &&
        state!.isNotEmpty &&
        state!.toLowerCase() != 'unknown') {
      return '$name, $state';
    }
    if (country.isEmpty || country.toLowerCase() == 'unknown') {
      return name;
    }
    return '$name, $country';
  }
}

extension StringExtension on String {
  String capitalize() {
    if (isEmpty) return this;
    return "${this[0].toUpperCase()}${substring(1)}";
  }
}
